 design/README.MD

```diff
@@ -0,0 +1,456 @@
+SdkInterface --> DafnyImplementation
+
+```
+
+The goal is for Dafny developers
+to call native AWS SDKs.
+
+### Wrapped Library
+
+`smithy-dafny` can wrap an existing `smithy-dafny` library.
+This is useful for integration testing.
+
+When you consider the first two use cases
+it is clear that you need tests
+to verify that these features are correct.
+
+The first thought is be to write tests in each runtime.
+But this is very expensive.
+Also keeping track of all these tests across different runtimes is difficult.
+You don't want to have tests only in some runtimes but not others.
+
+If you write the tests in Dafny then you only need to write the test once.
+But if you natively write these tests in Dafny
+when you execute them
+they will only test the Dafny code.
+Because they will not exercise the type conversions.
+
+So we want to write the tests in Dafny so that they are interoperable
+but we need to exercise the native code.
+By wrapping the whole project
+we force an additional type  conversion.
+
+Writing tests in Dafny means that
+your testing is now also interoperable.
+A Dafny tests can check multiple runtimes.
+But a simple test would only check the Dafny interfaces.
+
+For more details on how this works
+see the [polymorph development](#polymorph-development) section blow.
+
+This feature is used to test `smithy-dafny` on itself
+in the TestModels.
+This features is used by the Encryption SDK
+in the test vector runners
+to test interoperability between native runtimes.
+
+In this mode `smithy-dafny` will not produce Dafny types and interfaces.
+It will use the Dafny types and interfaces that already exist.
+It will also create a shim
+that will convert the Dafny to native
+and then call the native implementation.
+
+```mermaid
+---
+title: Wrapped Library
+---
+flowchart RL
+
+subgraph Wrapping[Wrapping Library]
+    direction RL
+    WrappingInterface[Generated Dafny SDK Interface] --> ToNativeW[Convert input \nto native]
+    ToDafnyW[Convert output \nto Dafny] --> WrappingInterface
+end
+
+subgraph Library[Native Library]
+    direction RL
+    Interface[Public Interface] --> ToDafny[Convert input \nto Dafny]
+    ToDafny --> DafnyImplementation[Dafny implementation]
+    DafnyImplementation --> ToNative[Convert output \nto native]
+    ToNative --> Interface
+end
+
+ToNativeW --> Interface
+Interface --> ToDafnyW
+
+DafnyTest --> WrappingInterface
+WrappingInterface --> DafnyTest
+
+```
+
+The goal is for Dafny developers
+to approximate native calls into Dafny libraries.
+`smithy-dafny` uses the wrapped library feature
+to test the correctness of polymorph-runtime features.
+
+Like this: `var output := ToDafny(ToNative(Work(ToDafny(ToNative(input)))))`
+
+## Polymorph development
+
+When developing `polymorph-X`
+the first thing to ask is:
+Does `smithy-X` exist?
+
+If it does exist then:
+`polymorph-X` SHOULD re-use as much `smithy-X` code-gen as possible.
+`polymorph-X` MUST use the same native types as `smithy-X`.
```

Comment:
Good to include at least one link to another `smithy-X` repo here as an example.